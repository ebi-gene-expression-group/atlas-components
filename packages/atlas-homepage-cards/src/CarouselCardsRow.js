import React from 'react'
import PropTypes from 'prop-types'
import styled, { createGlobalStyle } from 'styled-components'

import cardPropTypes from './modelPropTypes'
import Card from './Card'
import Slide from 'react-slick'

const CarouselCardsRow = (props) => {
  const { CardClass, cards, slideSettings } = props
  const { className, cardContainerClassName, speciesIconHeight, imageIconHeight, hoverColour } = props
  slideSettings.slidesToShow = slideSettings.slidesToShow >= cards.length ? cards.length : slideSettings.slidesToShow

  const CardContainer = styled.div`
    :hover {
      background: ${hoverColour};
    }
  `

  const SlideGlobalStyle = createGlobalStyle`
    .slick-slide img {
      margin: auto;
    }
    .slick-slider {
      margin: 30px auto 50px;
    }
    .slick-prev:before {
      color: #3497c5;
    }

    .slick-next:before{
      color: #3497c5;
    }
    .slick-prev:hover {
      color: #2f5767;
    }
    .slick-next:hover{
      color: #2f5767;
    }
`

  const cardsDisplay = Array.isArray(cards) && cards.map((card, index) =>
    <CardContainer className={cardContainerClassName} key={index}>
      <CardClass
        {...card}
        imageIconHeight={imageIconHeight}
        speciesIconHeight={speciesIconHeight}/>
    </CardContainer>)

  return (
    <div className={className}>
      <SlideGlobalStyle />
      {
        <Slide
          {...slideSettings}>
          {cardsDisplay}
        </Slide>
      }
    </div>
  )
}

CarouselCardsRow.propTypes = {
  cards: PropTypes.arrayOf(PropTypes.shape(cardPropTypes)).isRequired,
  CardClass: PropTypes.func,
  className: PropTypes.string,
  cardContainerClassName: PropTypes.string,
  speciesIconHeight: PropTypes.string,
  imageIconHeight: PropTypes.string,
  hoverColour: PropTypes.string,
  slideSettings: PropTypes.object
}

CarouselCardsRow.defaultProps = {
  CardClass: Card,
  className: ``,
  cardContainerClassName: ``,
  speciesIconHeight: `6rem`,
  imageIconHeight: `2rem`,
  hoverColour: `AliceBlue`,
  slideSettings: {
    dots: true,
    infinite: true,
    speed: 500,
    slidesToShow: 6,
    slidesToScroll: 1,
    autoplay: true,
    autoplaySpeed: 2000
  }
}

export default CarouselCardsRow
