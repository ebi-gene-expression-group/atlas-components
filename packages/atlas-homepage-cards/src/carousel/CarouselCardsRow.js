import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Slider from 'react-slick'

import cardPropTypes from '../modelPropTypes'
import Card from '../Card'

import { SlickStyle, SlickThemeStyle } from './SlickGlobalStyles'

const CarouselCardsRow = (props) => {
  const { CardClass, cards, sliderSettings, containerHeight, sliderHeight } = props
  const { className, cardContainerClassName, speciesIconHeight, imageIconHeight, hoverColour} = props

  const disableSlider = sliderSettings.slidesToShow >= cards.length
  sliderSettings.slidesToShow = disableSlider ? cards.length : sliderSettings.slidesToShow

  const CardContainer = styled.div`
    :hover {
      background: ${hoverColour};
    }
  `

  const cardsDisplay = Array.isArray(cards) && cards.map((card, index) =>
    <CardContainer className={cardContainerClassName} key={index}>
      <CardClass
        {...card}
        imageIconHeight={imageIconHeight}
        speciesIconHeight={speciesIconHeight}/>
    </CardContainer>)

  return (
    <div className={className} style={{height: containerHeight}}>
      {disableSlider ? cardsDisplay :
        <React.Fragment>
          <SlickStyle sliderHeight={sliderHeight}/>
          <SlickThemeStyle/>
          <Slider
            {...sliderSettings}>
            {cardsDisplay}
          </Slider>
        </React.Fragment>
      }
    </div>
  )
}

CarouselCardsRow.propTypes = {
  cards: PropTypes.arrayOf(PropTypes.shape(cardPropTypes)).isRequired,
  CardClass: PropTypes.func,
  className: PropTypes.string,
  cardContainerClassName: PropTypes.string,
  speciesIconHeight: PropTypes.string,
  imageIconHeight: PropTypes.string,
  hoverColour: PropTypes.string,
  sliderSettings: PropTypes.object.isRequired,
  //To prevent species summary panel's bottom border move up/down when tabs change.
  containerHeight: PropTypes.string,
  //To prevent slick-slider's bottom dots move up/down when tabs change as adaptiveHeight is applied.
  sliderHeight: PropTypes.string
}

CarouselCardsRow.defaultProps = {
  CardClass: Card,
  className: ``,
  cardContainerClassName: ``,
  speciesIconHeight: `6rem`,
  imageIconHeight: `2rem`,
  hoverColour: `AliceBlue`,
  containerHeight: ``,
  sliderHeight: ``
}

export default CarouselCardsRow
