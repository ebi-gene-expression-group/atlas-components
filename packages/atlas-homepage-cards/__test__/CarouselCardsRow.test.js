import { shallow, mount } from 'enzyme'
import Slide from 'react-slick'

import {
  aRickleInTimeImageCardProps, findingNemoSpeciesCardProps,     // URL in title, no URLs in content
  theSmithHouseholdImageCardProps, batmanFilmsSpeciesCardProps  // URLs in content, no URL in title
} from './TestUtils'

import CarouselCardsRow from '../src/CarouselCardsRow'

describe(`CarousleCardsRow`, () => {
  const props = {
    cards:
      [
        aRickleInTimeImageCardProps,
        batmanFilmsSpeciesCardProps,
        findingNemoSpeciesCardProps,
        theSmithHouseholdImageCardProps
      ],
    CardClass: () => `topClass`
  }

  test(`displays cards in slider mode if the number of cards is smaller than slides`, () => {
    const slideSettings = {slidesToShow: props.cards.length - 1}

    const wrapper = shallow(<CarouselCardsRow {...props} slideSettings={slideSettings}/>)
    expect(wrapper).toContainExactlyOneMatchingElement(Slide)
  })

  test(`displays cards in slider mode if the number of cards is larger than slides`, () => {
    const slideSettings = {slidesToShow: props.cards.length + 1}

    const wrapper = shallow(<CarouselCardsRow {...props} slideSettings={slideSettings}/>)
    expect(wrapper).not.toContainMatchingElement(Slide)
  })

  test(`displays all cards`, () => {
    expect(shallow(<CarouselCardsRow {...props}/>)).toContainMatchingElements(props.cards.length, props.CardClass);
  })

  test(`matches snapshot with displaying a part of cards in the slider`, () => {
    const slideSettings = {slidesToShow: props.cards.length - 1}

    expect(shallow(<CarouselCardsRow {...props} slideSettings={slideSettings}/>)).toMatchSnapshot()
  })

  test(`matches snapshot with displaying all cards in the slider`, () => {
    const slideSettings = {slidesToShow: props.cards.length + 1}

    expect(shallow(<CarouselCardsRow {...props} slideSettings={slideSettings}/>)).toMatchSnapshot()
  })
})
