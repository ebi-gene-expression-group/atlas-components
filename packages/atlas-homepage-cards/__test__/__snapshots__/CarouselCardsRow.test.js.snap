// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarousleCardsRow matches snapshot with displaying a part of cards in the slider 1`] = `
<div
  className=""
>
  <GlobalStyleComponent />
  <Slider
    slidesToShow={3}
  >
    <styled.div
      className=""
      key="0"
    >
      <CardClass
        content={
          Array [
            Object {
              "text": "Morty Smith",
            },
            Object {
              "text": "Summer Smith",
            },
            Object {
              "text": "Rick Sanchez",
            },
            Object {
              "text": "Rick Sanchez(split reality)",
            },
            Object {
              "text": "Morty Smith (split reality)",
            },
            Object {
              "text": "Summer Smith (split reality)",
            },
          ]
        }
        description={
          Object {
            "text": "A rickle in time",
            "url": "http://rickandmorty.wikia.com/wiki/A_Rickle_in_Time",
          }
        }
        iconSrc="https://upload.wikimedia.org/wikipedia/en/6/6b/A_Rickle_in_Time.jpg"
        iconType="image"
        imageIconHeight="2rem"
        speciesIconHeight="6rem"
      />
    </styled.div>
    <styled.div
      className=""
      key="1"
    >
      <CardClass
        content={
          Array [
            Object {
              "text": "Batman (1943)",
              "url": "https://en.wikipedia.org/wiki/Batman_(serial)",
            },
            Object {
              "text": "Batman and Robin (1949)",
              "url": "https://en.wikipedia.org/wiki/Batman_and_Robin_(serial)",
            },
            Object {
              "text": "Batman (1966)",
              "url": "https://en.wikipedia.org/wiki/Batman_(1966_film)",
            },
            Object {
              "text": "Batman (1989)",
              "url": "https://en.wikipedia.org/wiki/Batman_(1989_film)",
            },
            Object {
              "text": "Batman Returns (1992)",
              "url": "https://en.wikipedia.org/wiki/Batman_Returns",
            },
            Object {
              "text": "Batman Forever (1995)",
              "url": "https://en.wikipedia.org/wiki/Batman_Forever",
            },
            Object {
              "text": "Batman &amp; Robin (1997)",
              "url": "https://en.wikipedia.org/wiki/Batman_%26_Robin_(film)",
            },
            Object {
              "text": "Batman Begins (2005)",
              "url": "https://en.wikipedia.org/wiki/Batman_Begins",
            },
            Object {
              "text": "The Dark Knight (2008)",
              "url": "https://en.wikipedia.org/wiki/The_Dark_Knight_(film)",
            },
            Object {
              "text": "The Dark Knight Rises (2012)",
              "url": "https://en.wikipedia.org/wiki/The_Dark_Knight_Rises",
            },
            Object {
              "text": "Batman v Superman: Dawn of Justice (2016)",
              "url": "https://en.wikipedia.org/wiki/Batman_v_Superman:_Dawn_of_Justice",
            },
            Object {
              "text": "Suicide Squad (2016)",
              "url": "https://en.wikipedia.org/wiki/Suicide_Squad_(film)",
            },
            Object {
              "text": "Justice League (2017)",
              "url": "https://en.wikipedia.org/wiki/Justice_League_(film)",
            },
            Object {
              "text": "Joker (2019)",
              "url": "https://en.wikipedia.org/wiki/Joker_(2019_film)",
            },
          ]
        }
        description={
          Object {
            "text": "Batman films",
          }
        }
        iconSrc="bat"
        iconType="species"
        imageIconHeight="2rem"
        speciesIconHeight="6rem"
      />
    </styled.div>
    <styled.div
      className=""
      key="2"
    >
      <CardClass
        content={
          Array [
            Object {
              "text": "Dory",
            },
            Object {
              "text": "Nemo",
            },
            Object {
              "text": "Gill",
            },
          ]
        }
        description={
          Object {
            "text": "Finding Nemo",
            "url": "https://en.wikipedia.org/wiki/Finding_Nemo",
          }
        }
        iconSrc="zebrafish"
        iconType="species"
        imageIconHeight="2rem"
        speciesIconHeight="6rem"
      />
    </styled.div>
    <styled.div
      className=""
      key="3"
    >
      <CardClass
        content={
          Array [
            Object {
              "text": "Jerry Smith",
              "url": "https://en.wikipedia.org/wiki/List_of_Rick_and_Morty_characters#Jerry_Smith",
            },
            Object {
              "text": "Beth Smith",
              "url": "https://en.wikipedia.org/wiki/List_of_Rick_and_Morty_characters#Beth_Smith",
            },
            Object {
              "text": "Summer Smith",
              "url": "https://en.wikipedia.org/wiki/List_of_Rick_and_Morty_characters#Summer_Smith",
            },
            Object {
              "text": "Morty Smith",
              "url": "https://en.wikipedia.org/wiki/Morty_Smith",
            },
            Object {
              "text": "Rick Sanchez",
              "url": "https://en.wikipedia.org/wiki/Rick_Sanchez_(Rick_and_Morty)",
            },
          ]
        }
        description={
          Object {
            "text": "The Smith household",
          }
        }
        iconSrc="https://upload.wikimedia.org/wikipedia/en/b/b0/Rick_and_Morty_characters.jpg"
        iconType="image"
        imageIconHeight="2rem"
        speciesIconHeight="6rem"
      />
    </styled.div>
  </Slider>
</div>
`;

exports[`CarousleCardsRow matches snapshot with displaying all cards in the slider 1`] = `
<div
  className=""
>
  <styled.div
    className=""
    key="0"
  >
    <CardClass
      content={
        Array [
          Object {
            "text": "Morty Smith",
          },
          Object {
            "text": "Summer Smith",
          },
          Object {
            "text": "Rick Sanchez",
          },
          Object {
            "text": "Rick Sanchez(split reality)",
          },
          Object {
            "text": "Morty Smith (split reality)",
          },
          Object {
            "text": "Summer Smith (split reality)",
          },
        ]
      }
      description={
        Object {
          "text": "A rickle in time",
          "url": "http://rickandmorty.wikia.com/wiki/A_Rickle_in_Time",
        }
      }
      iconSrc="https://upload.wikimedia.org/wikipedia/en/6/6b/A_Rickle_in_Time.jpg"
      iconType="image"
      imageIconHeight="2rem"
      speciesIconHeight="6rem"
    />
  </styled.div>
  <styled.div
    className=""
    key="1"
  >
    <CardClass
      content={
        Array [
          Object {
            "text": "Batman (1943)",
            "url": "https://en.wikipedia.org/wiki/Batman_(serial)",
          },
          Object {
            "text": "Batman and Robin (1949)",
            "url": "https://en.wikipedia.org/wiki/Batman_and_Robin_(serial)",
          },
          Object {
            "text": "Batman (1966)",
            "url": "https://en.wikipedia.org/wiki/Batman_(1966_film)",
          },
          Object {
            "text": "Batman (1989)",
            "url": "https://en.wikipedia.org/wiki/Batman_(1989_film)",
          },
          Object {
            "text": "Batman Returns (1992)",
            "url": "https://en.wikipedia.org/wiki/Batman_Returns",
          },
          Object {
            "text": "Batman Forever (1995)",
            "url": "https://en.wikipedia.org/wiki/Batman_Forever",
          },
          Object {
            "text": "Batman &amp; Robin (1997)",
            "url": "https://en.wikipedia.org/wiki/Batman_%26_Robin_(film)",
          },
          Object {
            "text": "Batman Begins (2005)",
            "url": "https://en.wikipedia.org/wiki/Batman_Begins",
          },
          Object {
            "text": "The Dark Knight (2008)",
            "url": "https://en.wikipedia.org/wiki/The_Dark_Knight_(film)",
          },
          Object {
            "text": "The Dark Knight Rises (2012)",
            "url": "https://en.wikipedia.org/wiki/The_Dark_Knight_Rises",
          },
          Object {
            "text": "Batman v Superman: Dawn of Justice (2016)",
            "url": "https://en.wikipedia.org/wiki/Batman_v_Superman:_Dawn_of_Justice",
          },
          Object {
            "text": "Suicide Squad (2016)",
            "url": "https://en.wikipedia.org/wiki/Suicide_Squad_(film)",
          },
          Object {
            "text": "Justice League (2017)",
            "url": "https://en.wikipedia.org/wiki/Justice_League_(film)",
          },
          Object {
            "text": "Joker (2019)",
            "url": "https://en.wikipedia.org/wiki/Joker_(2019_film)",
          },
        ]
      }
      description={
        Object {
          "text": "Batman films",
        }
      }
      iconSrc="bat"
      iconType="species"
      imageIconHeight="2rem"
      speciesIconHeight="6rem"
    />
  </styled.div>
  <styled.div
    className=""
    key="2"
  >
    <CardClass
      content={
        Array [
          Object {
            "text": "Dory",
          },
          Object {
            "text": "Nemo",
          },
          Object {
            "text": "Gill",
          },
        ]
      }
      description={
        Object {
          "text": "Finding Nemo",
          "url": "https://en.wikipedia.org/wiki/Finding_Nemo",
        }
      }
      iconSrc="zebrafish"
      iconType="species"
      imageIconHeight="2rem"
      speciesIconHeight="6rem"
    />
  </styled.div>
  <styled.div
    className=""
    key="3"
  >
    <CardClass
      content={
        Array [
          Object {
            "text": "Jerry Smith",
            "url": "https://en.wikipedia.org/wiki/List_of_Rick_and_Morty_characters#Jerry_Smith",
          },
          Object {
            "text": "Beth Smith",
            "url": "https://en.wikipedia.org/wiki/List_of_Rick_and_Morty_characters#Beth_Smith",
          },
          Object {
            "text": "Summer Smith",
            "url": "https://en.wikipedia.org/wiki/List_of_Rick_and_Morty_characters#Summer_Smith",
          },
          Object {
            "text": "Morty Smith",
            "url": "https://en.wikipedia.org/wiki/Morty_Smith",
          },
          Object {
            "text": "Rick Sanchez",
            "url": "https://en.wikipedia.org/wiki/Rick_Sanchez_(Rick_and_Morty)",
          },
        ]
      }
      description={
        Object {
          "text": "The Smith household",
        }
      }
      iconSrc="https://upload.wikimedia.org/wikipedia/en/b/b0/Rick_and_Morty_characters.jpg"
      iconType="image"
      imageIconHeight="2rem"
      speciesIconHeight="6rem"
    />
  </styled.div>
</div>
`;
